### PR TITLE
Devendor some thridparty deps

### PR DESCRIPTION
**Pull request by Osyotr**

### CODE REVIEW

File: 3party/CMakeLists.txt [hunk 1]
```diff
@@ -41,6 +41,10 @@ if (NOT WITH_SYSTEM_PROVIDED_3PARTY)
 
   # Add protobuf library.
   add_subdirectory(protobuf)
+
+  add_library(utf8cpp INTERFACE)
+  add_library(utf8cpp::utf8cpp ALIAS utf8cpp)
+  target_include_directories(utf8cpp INTERFACE "${OMIM_ROOT}/3party/utfcpp/source")
```
Comment (biodranik) [inline 3party/CMakeLists.txt:55]: Why this change was required? Can it be extracted into a separate PR to test and polish properly?

File: CMakeLists.txt [hunk 2]
```diff
@@ -322,6 +322,7 @@ endif()
 if (WITH_SYSTEM_PROVIDED_3PARTY)
   set(GFLAGS_USE_TARGET_NAMESPACE ON)
   find_package(gflags REQUIRED)
+  find_package(utf8cpp REQUIRED)
```
Comment (biodranik) [inline CMakeLists.txt:324]: Is it really packaged on different systems?..

File: 3party/CMakeLists.txt [hunk 3]
```diff
@@ -22,6 +22,7 @@ if (NOT WITH_SYSTEM_PROVIDED_3PARTY)
   set(EXPAT_DTD OFF)
   set(EXPAT_NS ON)
   add_subdirectory(expat/expat)
+  add_library(expat::expat ALIAS expat)
```
Comment (biodranik) [inline 3party/CMakeLists.txt:25]: This change would also be great to test and merge separately.

File: CMakeLists.txt [hunk 4]
```diff
@@ -324,6 +324,7 @@ if (WITH_SYSTEM_PROVIDED_3PARTY)
   find_package(gflags REQUIRED)
 
   find_package(expat CONFIG REQUIRED)
+  find_package(jansson CONFIG REQUIRED)
```
Comment (biodranik) [inline CMakeLists.txt:322]: What is wrong with building these packages from sources?..

File: CMakeLists.txt [hunk 5]
```diff
@@ -320,6 +320,7 @@ if (WITH_SYSTEM_PROVIDED_3PARTY)
 
   find_package(expat CONFIG REQUIRED)
   find_package(jansson CONFIG REQUIRED)
+  find_package(pugixml REQUIRED)
```
Comment (biodranik) [inline CMakeLists.txt:323]: Why doesn't it compile from sources?

**Review by biodranik — COMMENTED** [review]
Thanks!

@Ferenc- can you please help with testing these changes?

File: transit/CMakeLists.txt [hunk 6]
```diff
@@ -20,7 +20,14 @@ set(SRC
 
 omim_add_library(${PROJECT_NAME} ${SRC})
 
-target_link_libraries(${PROJECT_NAME} PUBLIC jansson)
+target_link_libraries(${PROJECT_NAME}
+  PUBLIC
+    base
+    condig
```
Comment (Osyotr) [inline transit/CMakeLists.txt:26]:
```suggestion
    coding
```

File: cppjansson/CMakeLists.txt [hunk 7]
```diff
@@ -7,5 +7,7 @@ set(SRC
 )
 omim_add_library(${PROJECT_NAME} ${SRC})
 target_link_libraries(${PROJECT_NAME}
-  jansson
+  PUBLIC
```
[the remark this answers is not in the document]
Comment (Osyotr) [reply cppjansson/CMakeLists.txt:10]: Both are used in public interface.

File: CMakeLists.txt [hunk 8]
```diff
@@ -319,6 +319,11 @@ if (PLATFORM_LINUX)
   find_package(harfbuzz REQUIRED)
 endif()
 
+if (WITH_SYSTEM_PROVIDED_3PARTY)
```
[the remark this answers is not in the document]
Comment (Osyotr) [reply CMakeLists.txt:317]: Yes, but you need CMake 3.24+ for that
https://cmake.org/cmake/help/latest/variable/CMAKE_FIND_PACKAGE_TARGETS_GLOBAL.html#variable:CMAKE_FIND_PACKAGE_TARGETS_GLOBAL
Though IMHO it really shouldn't be in 3party.

File: 3party/CMakeLists.txt (hunk 1 above)
Comment (Osyotr) [reply 3party/CMakeLists.txt:55]: 1) Remove hardcoded `"${OMIM_ROOT}/3party/utf8cpp/include"` from include path
2) Unify usage for system-provided and vendored copy of utfcpp.

Changes in this PR somewhat depend on each other, I really wouldn't like to split them.

File: 3party/CMakeLists.txt (hunk 3 above)
Comment (Osyotr) [reply 3party/CMakeLists.txt:25]: Ditto wouldn't want to dive into rebase-cherry-pick-hell

File: CMakeLists.txt (hunk 2 above)
Comment (Osyotr) [reply CMakeLists.txt:324]: Those who set `WITH_SYSTEM_PROVIDED_3PARTY` should make sure they have it.
https://repology.org/project/utfcpp/versions

File: CMakeLists.txt (hunk 4 above)
Comment (Osyotr) [reply CMakeLists.txt:322]: Nothing in particular, though mixing system and vendored deps is considered bad.
Note that `add_subdirectory(jansson)` is not called on when `WITH_SYSTEM_PROVIDED_3PARTY`. CMake implicitly uses `-ljansson` if it couldn't find target with name `jansson`. It happens to work on Linux because everything is in include/link directories. This is why namespaced targets are preferred (they don't allow this implicit behavior).

File: CMakeLists.txt (hunk 5 above)
Comment (Osyotr) [reply CMakeLists.txt:323]: ditto

**Comment by Osyotr** [comment]
Please use squash when merging.

File: CMakeLists.txt [hunk 9]
```diff
@@ -103,7 +103,13 @@ if (NOT CMAKE_BUILD_TYPE)
 endif()
 
 # Global compile options for all configurations.
-add_compile_options(-ffast-math)
+if (MSVC)
+  add_compile_options(/utf-8)
+  add_link_options(/INCREMENTAL:NO)
+  add_link_options(/DEBUG:FULL)
```
[the remark this answers is not in the document]
Comment (Osyotr) [reply CMakeLists.txt:109]: Does it even make sense to have release build without debug symbols?

[the remark this answers is not in the document]
Comment (biodranik) [reply CMakeLists.txt:109]: Yes, when distributing a binary.

[the remark this answers is not in the document]
Comment (Osyotr) [reply CMakeLists.txt:109]: > Yes, when distributing a binary.

Windows binaries don't embed any debug information. I.e. without this flag you get `.exe`, with this flag you get `.exe` + `.pdb`.

[the remark this answers is not in the document]
Comment (biodranik) [reply CMakeLists.txt]: Does it influence the build speed?

[the remark this answers is not in the document]
Comment (Osyotr) [reply CMakeLists.txt]: My unscientific measurements (full build with tests, Release):
With flag: 8:50
Without flag: 8:30
Anyway, I've already removed it.

**Review by biodranik — APPROVED** [review]
Ok, let's move on with this to be able to run Windows soon. Thank you very much!

**Comment by Ferenc-** [comment]
> Thanks!
> 
> @Ferenc- can you please help with testing these changes?

So far, I haven't found any issues.
